### DISCUSSION TITLE
Add support for viewing and downloading stickers

### DISCUSSION THREAD

**Discussion by deralmas** [d99kris/nchat#95] [Ideas]
Hi! I really like this program and, although I don't know its architecture, a few things popped to mind.

Would it be possible to view the stickers? I can't seem to find a way. It looks like image stickers are WebP, video stickers are WebM and the animated vector ones are some weird https://github.com/TelegramMessenger/rlottie based thing.

All of them can be downloaded as files on the "regular" clients, so, unless this client archives them in some particular way, I think that it might be easy enough to make them viewable as regular files.

Again, mind that I have no actual idea on the feasibility of this proposal.

Another thing is that stickers have an emoji associated so I guess that a solution that's hopefully easier to implement would be to at least show the associated sticker next to the "sticker" text, kinda like how the "regular" clients do in their notifications, so that one could get at a glance what the sticker's probably going to express.

**Comment by d99kris** [comment]
Hi @Riteo - thanks for the suggestion, yes I agree it would be a useful feature. I think the emoji-equivalent text + ability to view the actual sticker should be possible without too much work. I'll create an "issue" to track the implementation of this. Will update here once added, probably this weekend or next.

**Reply by d99kris** [reply]
Issue #97 created.

**Comment by d99kris** [comment]
Hi again! Support for viewing / downloading stickers have been added now in 3a0e94d for Telegram and WhatsApp.

In Telegram the equivalent emoji e.g. `😄` is shown in the text message, for WhatsApp there is none so we just display `[Sticker]`. 

Selecting a message (key up/down) and using `CTRL-V` opens the sticker in system default viewer, `CTRL-R` downloads the sticker.

One may need to install a suitable image/video viewer that supports `webp`, `tgs` or any other file format used by Telegram / WhatsApp. I don't have any recommendation, but I used `feh` on Linux for viewing `webp` when testing out the sticker support.

**Reply by deralmas** [reply]
Wow, that was quick! Thanks a lot, the new feature works perfectly!

Note that I had to clear the history saved in `~/.nchat/history/Telegram_PHONENUMBER` in order to preview previously sent stickers properly. That makes sense, but I presume that other people stumbling on this feature might be a little bit confused, hope this helps.

**Reply by d99kris** [reply]
Ah, thanks for catching and pointing that out. 👍 

Yeah, my current approach when minor changes are made is to not bump the cache version to invalidate it. There are pros and cons to this. Perhaps I should add a simple command-line argument to nchat for clearing cache, and also try remember to point it out in feature/issue threads when changes are made. 🙂

**Reply by deralmas** [reply]
@d99kris yeah, that approach should improve usability in these cases. I'd still not exclude to people the idea of deleting the cache from the filesystem manually as it's a way more granular approach that basically comes for free.
